### ISSUE TITLE
Questionably constant?

### ISSUE THREAD

**Issue by eernstg** [dart-lang/language#4084]
The language team has had discussions about the need to change many parts of an expression in response to a small modification that makes a previously constant subexpression non-constant, or vice versa. For example:

```dart
class A {
  const A([_, _, _, _]);
}

int v = 1;

void main() {
  // Assume that we have this expression somewhere.
  const A(
    A(), 
    1, 
    'Hello', 
    A(<int>[], A(), A()),
  );

  // Now we need to change `[]` to `[v]`, where `v` isn't constant.
  A(
    const A(), 
    1, 
    'Hello', 
    A(<int>[v], const A(), const A()),
  );
}
```

This causes a lot of inconveniences in daily development because it's a non-trivial editing operation to make "as much as possible" constant in a given expression when we can't just say "all of it".

Arguably, the process is even more tricky in the other direction, because we can't see locally whether or not the list is expected to be modified in the future.

A more subtle semantic difference is the behavior of `identical`, where constant expressions yield canonicalized instances, which means that `const <int>[]` is the same object as another occurrence of the syntax `const <int>[]`, where as `<int>[]` and `<int>[]` are two distinct objects (assuming that they aren't constant based on a `const` modifier on some enclosing expression), even in the case where we are considering two distinct evaluations of the same expression in the source code.

## Proposal: `const?`

This issue is a proposal that we introduce the modifier `const?`, which would amount to a request that every subexpression of the expression that carries this modifier is constant, if possible.

The point is that it is much safer for an automatic mechanism to introduce `const` modifiers on a set of expressions if this operation is justified by an explicit request in the source code. We could say that the developer promises that it won't be a bug to implicitly add `const` to any of the subexpressions of the expression that has the modifier `const?`, nor will it be a bug if no such modifier is added.

The example above then becomes simpler: 

```dart
void main() {
  // Assume that we have this expression somewhere.
  const? A(
    A(), 
    1, 
    'Hello', 
    A(<int>[], A(), A()),
  );

  // Now we need to change `[]` to `[v]`, where `v` isn't constant.
  // Just go ahead and add `v`, no other change is needed.
  const? A(
    A(), 
    1, 
    'Hello', 
    A(<int>[v], A(), A()),
  );
}
```

This would largely eliminate the churn which is caused by changes to the constness of subexpressions of a bigger expression.

We could define the semantics of the `const?` modifier in terms of 'evidently constant expressions'. On the target expression, the mechanism would add `const` modifiers implicitly, in a bottom-up traversal, such that every expression gets the `const` modifier if and only if it is evidently constant.

- An expression is evidently constant if it has the modifier `const`. *For example, `const A()` or `const <int>[]`.*
- An identifier is evidently constant if and only if it is a constant expression (this is a reference to the rules that we have today). *For example, a name that denotes a constant variable or a top-level function.*
- An instance creation expression is evidently constant if every actual type argument is a constant type expression, and every actual argument is evidently constant. *For example, `C<int>(1)`.*
- A list literal is evidently constant if it has a type argument which is a constant type expression, or it has no actual type argument, and every element is evidently constant. Similarly for a set literal and a map literal.
- An operator expression (like `e1 + e2`, `!e3`, `e4 ? e5 : e6`) is evidently constant if the operands are evidently constant, and if it yields a constant expression when `const` is added (where it can be added syntactically) in a bottom-up traversal of the expression.

**Comment by abitofevrything** [comment]
Could this be expanded to allow `const?` constructors, e.g `const? Widget()`, such that any invocation of that constructor is made "as constant as possible"?

This would allow data classes that should never preoccupy themselves with canonicalization issues to advertise they should be const when possible, without the developer having to add `const` or `const?` to every instance creation expression.

Of course this then creates the issue that a constructor might be `const?`, but then a developer using that class might run into a use case where they need the instance _not_ to be const, so we'd need to introduce a way to explicitely disable the implicit const-ness.

**Comment by abitofevrything** [comment]
A user on discord (`@wdestroier`) has pointed out that `new` already provides a way to force the un-constedness of an expression.

**Comment by eernstg** [comment]
> Could this be expanded to allow `const?` constructors, e.g `const? Widget()`, such that any invocation of that constructor is made "as constant as possible"?

That's an interesting idea! It might be possible. It might also be a bit too implicit (in the sense that we'll end up having some lists that are constant "by accident", even though they must actually be mutated at some point). But it's certainly worth keeping in mind.

> ... need to introduce a way to explicitely disable ...

Indeed, that's another reason why it might be a bit too aggressive to have a `const?` constructor.

**Comment by lrhn** [comment]
Can you add `const?` only where you can write `const` today, or can you write `const? (1 + 2)` too?
Can you use it on the LHS of a declaration, `const? c = ...;`?

I'll assume it's only a replacement for `const`, so before an object creation or collection literal expression.

That would put the reast of that expression in a `const?` context, like `const` puts it in a `const` context.

Then we use the same rules as for constant expressions today, with small twists:
* If it's not one of the expression contexts which can be constant, it's not constant.
* Otherwise if it's an object creation or collection literal prefixed by `const` then that introduces a `const` context on subexpressions.
* Otherwise the subexpressions are in the same `const` or `const?` context as the parent expression, and we recurse as normal.
* When we come back, then for every syntactic form we use the same rules as today for deciding whether the expression is constant, but if it's in a `const?` context instead of a `const` context, it's not an error to not be constant.

So effectively we pass down a flag on the `const` context that says whether it's going to be an error to not be `const`.
If any required expression ends up not being constant, and therefore not having a value at compile time, then so does the expression containing it.

There are expressions where a sub-expression doesn't have to be constant, if it's in a branch not taken.
```dart
const c = true ? "banana" : (1 ~/ 0);
```
Here `(1 ~/ 0)` is not a constant expression because it doesn't evaluate to a value. It throws. But it's also not used.

If we do:
```dart
var v = const? [bool.hasEnvironment("override") ? int.fromEnvironment("override") : computeValue()];
```
then the expression is not "syntactically a constant expression", which is a bad concept anyway.
It should be equivalent to:
```dart
var v = [const bool.hasEnvironment("override") ? const int.fromEnvironment("override") : computeValue()];
```

I think this can be built into the existing rules with fairly little work. The biggest issue would be existing assumptions and bad definitions (and the lack of actually defining "constant evaluation" and when it happens, making `var x = "a" + "b";`'s initializer expression be not in a constant context, but "is a constant expression". Is it *evaluated* as a constant? That's the question we need to answer.).

**Comment by lrhn** [comment]
Can you add `const?` only where you can write `const` today, or can you write `const? (1 + 2)` too?
Can you use it on the LHS of a declaration, `const? c = ...;`?

I'll assume it's only a replacement for `const`, so before an object creation or collection literal expression.

That would put the reast of that expression in a `const?` context, like `const` puts it in a `const` context.

Then we use the same rules as for constant expressions today, with small twists:
* If it's not one of the expression contexts which can be constant, it's not constant.
* Otherwise if it's an object creation or collection literal prefixed by `const` then that introduces a `const` context on subexpressions.
* Otherwise the subexpressions are in the same `const` or `const?` context as the parent expression, and we recurse as normal.
* When we come back, then for every syntactic form we use the same rules as today for deciding whether the expression is constant, but if it's in a `const?` context instead of a `const` context, it's not an error to not be constant.

So effectively we pass down a flag on the `const` context that says whether it's going to be an error to not be `const`.
If any required expression ends up not being constant, and therefore not having a value at compile time, then so does the expression containing it.

There are expressions where a sub-expression doesn't have to be constant, if it's in a branch not taken.
```dart
const c = true ? "banana" : (1 ~/ 0);
```
Here `(1 ~/ 0)` is not a constant expression because it doesn't evaluate to a value. It throws. But it's also not used.

If we do:
```dart
var v = const? [bool.hasEnvironment("override") ? int.fromEnvironment("override") : computeValue()];
```
then the expression is not "syntactically a constant expression", which is a bad concept anyway.
It should be equivalent to:
```dart
var v = [const bool.hasEnvironment("override") ? const int.fromEnvironment("override") : computeValue()];
```

I think this can be built into the existing rules with fairly little work. The biggest issue would be existing assumptions and bad definitions (and the lack of actually defining "constant evaluation" and when it happens, making `var x = "a" + "b";`'s initializer expression be not in a constant context, but "is a constant expression". Is it *evaluated* as a constant? That's the question we need to answer.).

**Comment by eernstg** [comment]
> Can you add `const?` only where you can write `const` today,

That's my starting point, but if it turns out to be useful to allow it in additional positions then we can scrutinize the grammar and see if it can be done. Allowing `const? (<expression>)` is probably easy to handle in the grammar.

> Can you use it on the LHS of a declaration, `const? c = ...;`?

That seems less intuitive to me. I'm thinking of `const?` as an operator that works on a structure, not a property of the variable. For example, we might want to make the initializing expression as constant as possible, and still allow the variable to be mutable or not, and late or not. In that situation `const?` seems to be noise more than signal, if it is intended to have an effect on the initializing expression and not on the variable.

A '`const?` context' is a nice way to describe it. I'm not sure about the distinction (it might be completely equivalent to the notion of 'evidently constant expressions' that I mentioned), or about pros/cons in this choice.

**Comment by abitofevrything** [comment]
> (in the sense that we'll end up having some lists that are constant "by accident", even though they must actually be mutated at some point

Simple enough: don't make the list constructors `const?`.

I would see `const?` constructors as more of a feature for custom dataclasses, e.g classes generated by freezed, or maybe potentially widget classes.

**Comment by cedvdb** [comment]
This is a nice solution for something that's inconvenient

It looks a bit like a self inflicted wound to me. I'm wondering if pushing `const` so much (via the analyzer, prefer const) was the way to go. I've yet to see a realistic benchmark that a flutter project using const has a non negligile performance boost (the PR with a benchmark on the official flutter repo is not a real world scenario). Why is it pushed forward so much given the inconveniences above ? 

Maybe const by default was worth consideration too, people can figure out equality issues ? I don't know.

**Comment by mateusfccp** [comment]
> I'm also wondering if pushing `const` so much (via the analyzer, prefer const) was the way to go. I've yet to see a realistic benchmark that a flutter project using const has a non negligile performance boost.

It seems there are many people who don't like `const`ing the code. You are not the first one I've seen complaining about this linter, and many even disable it and simply don't put `const` in front of their `const`able objects.

Particularly, I think it's a costless optimization. Even if it may be negligible for many cases, it's basically costless, except for a little bit of verbosity, and this is exactly what this issue proposes to solve.

There's an alternative approach. We could make something similar to what was proposed in #1410, but instead of making everything `const`, make everything `const?`, according to the semantics defined in this issue.

Then, we would have to revert `const`-ness to be opt-out instead of opt-in. That is, whenever we don't want a `const`able object creation to be `const`, we would have to use `new`.

Is this better than what we currently have? I'm not sure. We would have to extract some data to have a better idea which case is more common. In my personal experience with Flutter, we would have much less `new` than we have `const` today (considering that we `const` everything that should be `const`). I have no extensive experience with non-Flutter Dart to have any impression on it.

However, even if the data says that this is also true for non-Flutter Dart, it is an extremely breaking change to introduce in the language now.

**Comment by cedvdb** [comment]
@mateusfccp I wondered about that too

`const?` opens the door to `const?` by default because it does not suffer from the same drawbacks than `const` by default.

First, if you want errors if an expression is not a const, you can use `const`.

But it seem unrealistic because of equality issues in existing code bases.

>  I think it's a costless optimization

Well it's not costless at the moment, that's why this proposal exists

**Comment by lucavenir** [comment]
I like the fact that there's `const?` and `const`. The former helps with an optimization, the latter is an explicit take the developer chooses to follow.

Still, I'm not sure I've understood how this proposal helps with declarations. That's half of the times I encounter the "churn" problem. Example:

```dart
class Point {
  const Point(this.x, this.y);
  final double x;
  final double y;
}

const x = 2.0;
const y = 3.0;

// can be const, so the auto fix will eventually make this `const` on save
final p = Point(2, 3);
```

At this point I save and the lint successfully turns `p` into `const p`.
Then, I need to change `y` for some reason.

```dart
import 'dart:math';
// ...

const x = 2.0;
const y = Random().nextDouble();

const p = Point(2, 3);  // error
```

So `const?` would help even on the LHS.
The difference with `const`, at that point, would be explicitness.

In other words, if I understood this proposal correctly, when performance is a must, the developer must avoid `const?`.

**Comment by eernstg** [comment]
@lucavenir, did you mean `Point(x, y)` rather than `Point(2, 3)`? Also, `final y` rather than `const y` in the second version? Otherwise, I can't see why there would be an error at the declaration of `p`.

You could use the following if you want the best performance, and you don't want to touch the declaration of `p`  in response to a change to the constness of `x` and/or `y`:

```dart
final p = const? Point(x, y);
```

This would make `p` final, but not constant, which means that it cannot be used in constant expressions. However, `p` will be initialized with the value of a constant expression if and only if both x and y are constant doubles. This means that `p` will be _just as good as a constant variable_ when it is used for any other purpose than being a subexpression of a constant expression (where it's an error).

Of course, you could also declare `p` as `const p = const? Point(x, y);` but that's silly because that occurrence of `const?` can never make a difference: if it doesn't simply mean `const Point(x, y)`, it's an error because we have `const p`, and `const Point(x, y)` is indeed the meaning of `Point(x, y)` when `p` is constant.

> when performance is a must, the developer must avoid `const?`.

Well, if you insist that a given expression must be a constant expression then you must use `const` rather than `const?`.

But otherwise it's the other way around: You would use `const?` in a large number of locations in order to make the target expression a constant expression _as far as possible_, which would presumably ensure that your program has a larger number of constant expressions than it would otherwise have had (because we don't like the churn, so we don't add `const` meticulously in every single location that allows it).

**Comment by lucavenir** [comment]
> @lucavenir, did you mean `Point(x, y)` rather than `Point(2, 3)`? Also, `final y` rather than `const y` in the second version? Otherwise, I can't see why there would be an error at the declaration of `p`.

Yes, sorry! I'll edit my comment to make it clearer.

**Comment by rakudrama** [comment]
I would hope that a `const?` expression modifier as proposed would be extended to declaring `const?` constructors.
```dart
  const? Point(this.x, this.y);
```

For the use-case of Flutter widget trees, framework users would almost never have to write `const` _or_ `const?`.

See https://github.com/dart-lang/language/issues/3399

**Comment by Levi-Lesches** [comment]
After all this discussion, I'm still not sure I see the benefits of `const?`. 
- It makes more declarations about const-ness implicit, which isn't necessarily good
- It hides whether the objects are actually const, forcing you to either notice or use the standard form, `const`
- It adds another option, "plain", `const`, or `const?` to the already-confusing concept of const-ness
- It doesn't solve the case of having a non-const object deep inside a `const?` object
- I can see it becoming encouraged as "use `const?` when you don't know what to do", and then objects suddenly become const, which will trip up beginners because they can't see what is and what isn't const
- Judging by the conversation here, it's not necessarily clear to everyone what the exact semantics of `const?` are or would be. 

Given that the original "problem" was the chore of needing to add `const` everywhere, I would agree with [this comment](https://github.com/dart-lang/language/issues/3399#issuecomment-1792524799) and say that running `dart fix` is quick enough to make this a moot point. As an advantage, developers are made aware of when objects can now or no longer be const, and get to make those choices for themselves, while preserving readability in the future.

**Comment by cedvdb** [comment]
> Given that the original "problem" was the chore of needing to add const everywhere, I would agree with https://github.com/dart-lang/language/issues/3399#issuecomment-1792524799 and say that running dart fix is quick enough to make this a moot point. As an advantage, developers are made aware of when objects can now or no longer be const, and get to make those choices for themselves, while preserving readability in the future.

I somewhat disagree with this. 

  - Often times I spin up a test project just to try things out and I have to deal with const / unconsting stuff. When I'm just prototyping this gets in the way.
  - The fact that prefer const is used by default is weird, most projects don't reach the production phase, let alone a squeeze as much perf as you can phase. 
  - Having the option to have the *potential* performance gains and not having to care about the above would be nice. Ultimately as an end user a mechanism that wouldn't necessitate an additional marker would be nicer though.

**Comment by eernstg** [comment]
I adjusted the proposal to include `const?` constructors. I'm a little worried about their ability to introduce subtle bugs at a grandiose scale, but it might be OK when we get used to them. :grinning:

**Comment by cedvdb** [comment]
@eernstg 

about `const?` constructors, which look nice on the surface

  > This implies that it can be invoked with the const modifier,

It would be more useful if a `const?` constructor may or may not be invoked with the const modifier. That is, `const?` constructor would be allowed on constructors that cannot be const.

Otherwise in flutter, in the build method, `return const?` will still have to be used unless you know for a fact that the constructor of the top level widget is `const?`, which would be painful to check (and maintain) on a case by case basis.

The implications are the same anyway, concerning potential bugs this could introduce.  

 > I'm a little worried about their ability to introduce subtle bugs at a grandiose scale, but it might be OK when we get used to them. 

This looks more dangerous than if Dart was designed from the get go with a `noconst` modifier and `const?` by default.  That would just be a feature of the language, easy to reason about, here it's feature of the constructor. Personally, I'm not comfortable with it.

**Comment by eernstg** [comment]
> It would be more useful if a `const?` constructor would be allowed on constructors that cannot be `const`.

Oh, that's a great point! So we'd want to express the rule that all subexpressions of a given constructor invocation should be treated as if they were in scope of a `const?` expression modifier, even though we already have a guarantee that we can't make the outermost expression `const`.

One gnarly problem is the syntax: I to think that a constructor declaration like `const? const A();` is intolerably keyword heavy.

We could also consider having support for a `const?` modifier on a formal parameter declaration. This would ensure that every actual argument `a` passed to that parameter would be treated as `const? a`.

We would then be able to express the behavior that you mention by adding `const?` on all formal parameters of the constructor which should otherwise have the `const?` modifier even though it can't be `const` itself (so we avoid the contradiction, and we avoid the bulky use of `const? const`).

**Comment by eernstg** [comment]
I added a proposal for `const?` parameters as well.
